### ISSUE TITLE
Messaging: Generic Sorted Messages

### ISSUE THREAD

**Issue by ptheywood** [FLAMEGPU/FLAMEGPU2_dev#82]
Messages sorted by a specific key, either sorting all possible values or optimised to an upper limit of bins.

Could be specialised for special cases :
 
+ 1 message per key
+ along with multiple messages per key

Access would then be based on the user requesting messages from a specific key.

**Comment by Robadob** [comment]
* `MsgArray`
    * 1 message per key (enforced with atomics?)
    * Keys form a contiguous integer block 
    * User sets min/max index* `ArrayMsg`
    * Could add a convenience function for interpolating messages variables? (for 2D/3D might be possible to back this with texture memory intrinsics.
* `MsgArray2D`
    * Same as above, 2D indexes, provides classic FGPU discrete functionality
    * Maybe a Moore neighbourhood iteration option.
* `MsgArray3D`
    * Same as above, 3D indexes
    * Maybe a Moore neighbourhood iteration option.

Little unsure about these ones:

* `MapMsg`:
   * Msg per key
   * User defined hash function
* `MultiMapMsg`: 
    * Multiple messages per key
    * User provides a hash function?
    * Question is over whether we would want it to be a real hash map, whereby we don't return messages with different key but same hash.
    * Basically has Spatial PBM at back end, could possibly rewrite spatial messages to inherit from

**Comment by mondus** [comment]
Don't call it map call it bucket?
